src/realm/list.hpp

```diff
@@ -163,85 +215,146 @@ template <>
 void Lst<ObjLink>::do_remove(size_t);
 extern template class Lst<ObjLink>;
 
-class LnkLst : public Lst<ObjKey>, public ObjList {
+// Extern template declarations for lists of primitives:
+extern template class Lst<int64_t>;
+extern template class Lst<bool>;
+extern template class Lst<StringData>;
+extern template class Lst<BinaryData>;
+extern template class Lst<Timestamp>;
+extern template class Lst<float>;
+extern template class Lst<double>;
+extern template class Lst<Decimal128>;
+extern template class Lst<ObjectId>;
+extern template class Lst<UUID>;
+extern template class Lst<util::Optional<int64_t>>;
+extern template class Lst<util::Optional<bool>>;
+extern template class Lst<util::Optional<float>>;
+extern template class Lst<util::Optional<double>>;
+extern template class Lst<util::Optional<ObjectId>>;
+extern template class Lst<util::Optional<UUID>>;
+
+class LnkLst final : public ObjCollectionBase<LstBase> {
 public:
+    using Base = ObjCollectionBase<LstBase>;
+    using value_type = ObjKey;
+    using iterator = CollectionIterator<LnkLst>;
+
     LnkLst() = default;
 
-    LnkLst(const Obj& owner, ColKey col_key);
-    LnkLst(const LnkLst& other)
-        : Lst<ObjKey>(other)
-        , m_unresolved(other.m_unresolved)
+    LnkLst(const Obj& owner, ColKey col_key)
+        : m_list(owner, col_key)
     {
+        update_unresolved();
     }
-    LnkLst& operator=(const LnkLst& other)
+
+    LnkLst(const LnkLst& other) = default;
+    LnkLst(LnkLst&& other) = default;
+    LnkLst& operator=(const LnkLst& other) = default;
+    LnkLst& operator=(LnkLst&& other) = default;
+    bool operator==(const LnkLst& other) const;
+    bool operator!=(const LnkLst& other) const;
+
+    Obj operator[](size_t ndx) const
     {
-        Lst<ObjKey>::operator=(other);
-        m_unresolved = other.m_unresolved;
-        return *this;
+        return get_object(ndx);
     }
 
-    LnkLstPtr clone() const
+    ObjKey get(size_t ndx) const;
+    size_t find_first(const ObjKey&) const;
+    void insert(size_t ndx, ObjKey value);
+    ObjKey set(size_t ndx, ObjKey value);
+    ObjKey remove(size_t ndx);
+
+    void add(ObjKey value)
     {
-        if (m_obj.is_valid()) {
-            return std::make_unique<LnkLst>(m_obj, m_col_key);
+        // FIXME: Should this add to the end of the unresolved list?
+        insert(size(), value);
+    }
+
+    // Overriding members of CollectionBase:
```

Comment:
I don't think we have this as part of the style guide or anything, but why not make all these methods use the override keyword so the compiler can help us verify that these actually do override something?

Reply:
Thanks for this question, I learned something looking up the following:

`final` has the same effect (i.e., causes a compiler error if it doesn't actually override anything in this case). `override final` only really does anything if the `virtual` keyword is also present in the declaration, where `final` alone does not imply `override`.

So:

```
// These are identical:
void foo() override final;
void foo() final;

// No requirement that anything is overridden from base class, but prevents override in derived classes:
virtual void foo() final;

// Requires that something is overridden:
virtual void foo() override final;
```

The use case for `virtual void foo() final` is to prevent _static_ override, which could actually be useful for something else.

Reply:
Ok great, I'm happy to learn that using final is sufficient for this!